### PR TITLE
`Paywalls`: new `onRestoreCompleted` handler

### PR DESCRIPTION
**Pull request by NachoSoto**
This is available on `.paywallFooter()`, `.presentPaywallIfNeeded`, and as a `.onRestoreCompleted` modifier.

### CODE REVIEW

File: RevenueCatUI/View+PresentPaywall.swift [hunk 1]
```diff
@@ -144,6 +154,13 @@ private struct PresentingPaywallModifier: ViewModifier {
 
                         self.close()
                     }
+                    .onPurchaseCompleted { customerInfo in
+                        self.restoreCompleted?(customerInfo)
+
+                        if !self.shouldDisplay(customerInfo) {
+                            self.close()
+                        }
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywall.swift:162]: Data driven UI FTW: we can make it dismiss automatically as well if the condition becomes `false` after restoring 🎉

File: RevenueCatUI/View+PurchaseRestoreCompleted.swift [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    ///  }
+    /// ```
+    ///
+    /// - Warning: Receiving a ``CustomerInfo``does not imply that the user has any entitlements,
+    /// simply that the process was successful. You must verify the ``CustomerInfo/entitlements``
+    /// to confirm that they are active.
+    ///
+    /// ### Related Articles
+    /// [Documentation](https://rev.cat/paywalls)
+    public func onRestoreCompleted(
```
Comment (tonidero) [inline RevenueCatUI/View+PurchaseRestoreCompleted.swift:77]: I wonder, would it make sense to group all these callbacks together? In android, I was planning to create a "Listener" interface, that can be implemented by the developer to listen to any callbacks they need to.

File: RevenueCatUI/View+PresentPaywall.swift [hunk 3]
```diff
@@ -144,6 +154,13 @@ private struct PresentingPaywallModifier: ViewModifier {
 
                         self.close()
                     }
+                    .onPurchaseCompleted { customerInfo in
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywall.swift:157]:
```suggestion
                    .onRestoreCompleted { customerInfo in
```

**Review by joshdholtz — APPROVED** [review]
YESSSSSS 🥳

File: RevenueCatUI/View+PurchaseRestoreCompleted.swift (hunk 2 above)
Comment (NachoSoto) [reply RevenueCatUI/View+PurchaseRestoreCompleted.swift:77]: That would be the `UIKit` way, with a delegate `protocol`. But no in SwiftUI...

File: RevenueCatUI/UIKit/PaywallViewController.swift [hunk 4]
```diff
@@ -76,8 +80,17 @@ public protocol PaywallViewControllerDelegate: AnyObject {
...
+    optional func paywallViewController(_ controller: PaywallViewController,
+                                        didFinishPurchasingWith customerInfo: CustomerInfo)
+
+    /// Notifies that the restore operation has completed in a ``PaywallViewController``.
+    ///
+    /// - Warning: Receiving a ``CustomerInfo``does not imply that the user has any entitlements,
+    /// simply that the process was successful. You must verify the ``CustomerInfo/entitlements``
+    /// to confirm that they are active.
+    @objc(paywallViewController:didFinishRestoringWithCustomerInfo:)
+    optional func paywallViewController(_ controller: PaywallViewController,
```
Comment (NachoSoto) [inline RevenueCatUI/UIKit/PaywallViewController.swift:92]: Added this too just for @codykerns

**Pull request merged**